[quantization] Re-write of tensor-scalar mul

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26937 [quantization] Re-write of tensor-scalar mul**

Differential Revision: [D17618028](https://our.internmc.facebook.com/intern/diff/D17618028)

Commit:
[quantization] Re-write of tensor-scalar mul

raghuramank100 commented on aten/src/ATen/native/quantized/cpu/qmul.cpp:
```diff
@@ -50,22 +51,58 @@ Tensor _mul_out(Tensor& out, const Tensor& self, const Tensor& other) {
+    int64_t q_max = std::numeric_limits<underlying_t>::max();
+
+    if (other_val > 0.0) {
+      scale_prime = other_val * self_scale;
```
Where is the code for handling other_val < 0?

raghuramank100 commented on aten/src/ATen/native/quantized/cpu/qmul.cpp:
```diff
@@ -50,22 +51,58 @@ Tensor _mul_out(Tensor& out, const Tensor& self, const Tensor& other) {
+          scale_prime, zero_point_prime, self.scalar_type()));
+    } else /* other_val < 0.0 */ {
+      scale_prime = other_val * self_scale;
+      zero_point_prime = q_max - (self_zero_point - q_min) - 1;
```
zero_point_prime should be = q_max -(self_zero_point - q_min). There should be no -1.

raghuramank100 commented on test/test_quantized.py:
```diff
@@ -331,11 +331,9 @@ def test_qmul_relu_same_qparams(self):
-            self.assertEqual(dqC, dqC_hat)
+            qC_hat = mul(qA, b.item())
+
+            self.assertEqual(C_ref, qC_hat.dequantize())
```
This test should not be passing with the existing code for negative b. If it does, then we are missing something. Can you check?

Commit:
Update on "[quantization] Re-write of tensor-scalar mul"


Differential Revision: [D17618028](https://our.internmc.facebook.com/intern/diff/D17618028)

Commit:
Update on "[quantization] Re-write of tensor-scalar mul"


Differential Revision: [D17618028](https://our.internmc.facebook.com/intern/diff/D17618028)

raghuramank100 commented on test/test_quantized.py:
```diff
@@ -328,14 +328,19 @@ def test_qmul_relu_same_qparams(self):
+
+            self.assertEqual(C_ref, qC_hat.dequantize())
+
+        # Scalar addition + relu
```
Looks good. Minor NIT: Fix comments to say multiplication here and at L 330.

Commit:
Update on "[quantization] Re-write of tensor-scalar mul"


Differential Revision: [D17618028](https://our.internmc.facebook.com/intern/diff/D17618028)
